shorebird_code_push/lib/src/shorebird_updater_io.dart

```diff
@@ -0,0 +1,159 @@
+      _updater.currentPatchNumber();
+      _isAvailable = true;
+    } catch (_) {
+      logShorebirdEngineUnavailableMessage();
+      _isAvailable = false;
+    }
+  }
+
+  late final bool _isAvailable;
+
+  final Updater _updater;
+
+  final IsolateRun _run;
+
+  @override
+  bool get isAvailable => _isAvailable;
+
+  @override
+  Future<Patch?> readCurrentPatch() async {
+    if (!_isAvailable) return null;
+
+    return _run(
+      () {
+        try {
+          final patchNumber = _updater.currentPatchNumber();
+          return patchNumber > 0 ? Patch(number: patchNumber) : null;
+        } catch (error) {
+          throw ReadPatchException(message: '$error');
+        }
+      },
+    );
+  }
+
+  @override
+  Future<Patch?> readNextPatch() async {
+    if (!_isAvailable) return null;
+
+    return _run(
+      () {
+        try {
+          final patchNumber = _updater.nextPatchNumber();
+          return patchNumber > 0 ? Patch(number: patchNumber) : null;
+        } catch (error) {
+          throw ReadPatchException(message: '$error');
+        }
+      },
+    );
+  }
+
+  @override
+  Future<UpdateStatus> checkForUpdate() async {
+    if (!_isAvailable) return UpdateStatus.unavailable;
+
+    final isUpdateAvailable = await _run(_updater.checkForUpdate);
+    if (isUpdateAvailable) return UpdateStatus.outdated;
+
+    final (current, next) = await (readCurrentPatch(), readNextPatch()).wait;
+    return next != null && current?.number != next.number
+        ? UpdateStatus.restartRequired
+        : UpdateStatus.upToDate;
+  }
+
+  @override
+  Future<void> update() async {
+    if (!_isAvailable) return;
+
+    late final Pointer<UpdateResult> result;
+
+    try {
+      result = await _run(_updater.downloadUpdateWithError);
+
+      if (result == nullptr) {
+        const reason = UpdateFailureReason.unknown;
+        final message = reason.toFailureMessage();
+        throw UpdateException(reason: reason, message: message);
+      }
+
+      final status = result.ref.status;
+
+      if (status == 1) return; // SHOREBIRD_UPDATE_SUCCESS
+
+      final reason = status.toFailureReason();
+      final details = result.ref.message != nullptr
+          ? result.ref.message.cast<Utf8>().toDartString()
+          : 'unknown';
+      final message = reason.toFailureMessage(details);
+      throw UpdateException(message: message, reason: reason);
+    } finally {
+      _updater.freeUpdateResult(result);
+    }
+  }
+}
+
+extension on int {
+  UpdateFailureReason toFailureReason() {
+    switch (this) {
+      case 0: // SHOREBIRD_NO_UPDATE
+        return UpdateFailureReason.noUpdate;
+      case 2: // SHOREBIRD_UPDATE_HAD_ERROR
+        return UpdateFailureReason.downloadFailed;
+      case 3: // SHOREBIRD_UPDATE_IS_BAD_PATCH
+        return UpdateFailureReason.badPatch;
+      case 4: // SHOREBIRD_UPDATE_ERROR
+        return UpdateFailureReason.unknown;
+      default:
+        return UpdateFailureReason.unknown;
```

Comment:
I think we can find a way to share these constants. I'm surprised the ffi gen didn't do it for you.

Reply:
Do we want to directly export the constants from ffigen? They aren't very dart-like so I opted to transform them to more darty enums with API docs etc.

Reply:
I don't really care one way or another.  I had just assumed that ffigen would have picked them up and thus allowing us to not have to worry about keeping two codebases in sync.

The FFI layer is already non-darty, it's an implementation detail of package:shorebird_code_push.

Reply:
The `UpdateFailureReason` is a public part of the package:shorebird_code_push API so it's not an implementation detail.

Reply:
Great.  Yeah, none of the rust/C api is public, and it makes sense to have a translation layer.